session/filetransfer.go

```diff
@@ -0,0 +1,496 @@
+		} else {
+			log.Errorf("Received error from client: %s", *erro.Error)
+		}
+
+	default:
+		h.Error(msg, w, errors.Errorf(
+			"session: filetransfer message type '%s' not supported",
+			msg.Header.MsgType,
+		))
+	}
+}
+
+func (h *FileTransferHandler) InitFileDownload(msg *ws.ProtoMsg, w ResponseWriter) (err error) {
+	params := new(GetFile)
+	if err := msgpack.Unmarshal(msg.Body, params); err != nil {
+		err = errors.Wrap(err, "malformed request parameters")
+		h.Error(msg, w, err)
+		return err
+	} else if err = params.Validate(); err != nil {
+		err = errors.Wrap(err, "malformed request parameters")
+		h.Error(msg, w, err)
+		return err
+	}
+	select {
+	case h.mutex <- struct{}{}:
+		defer func() {
+			if err != nil {
+				<-h.mutex
+			}
+		}()
+	default:
+		err = errors.New("another file transfer is still in progress")
+		h.Error(msg, w, err)
+		return err
+	}
+	fd, err := os.Open(*params.Path)
+	if err != nil {
+		err = errors.Wrap(err, "failed to open file for reading")
+		h.Error(msg, w, err)
+		return err
+	}
+	go h.DownloadHandler(fd, msg, w)
+	return nil
+}
+
+type chunkWriter struct {
+	SessionID string
+	Offset    int64
+	W         ResponseWriter
+}
+
+func (c *chunkWriter) Write(b []byte) (int, error) {
+	msg := ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeFileTransfer,
+			MsgType:   wsft.MessageTypeChunk,
+			SessionID: c.SessionID,
+			Properties: map[string]interface{}{
+				"offset": c.Offset,
+			},
+		},
+		Body: b,
+	}
+	err := c.W.WriteProtoMsg(&msg)
+	if err != nil {
+		return 0, err
+	}
+	c.Offset += int64(len(b))
+	return len(b), err
+}
+
+func (h *FileTransferHandler) DownloadHandler(fd *os.File, msg *ws.ProtoMsg, w ResponseWriter) (err error) {
+	const (
+		BufSize = 4096
+	)
+	defer func() {
+		<-h.mutex
+		errClose := fd.Close()
+		if errClose != nil {
+			log.Warn("error closing file descriptor: %s", errClose.Error())
+		}
+		if err != nil {
+			h.Error(msg, w, err)
+			log.Error(err.Error())
+		}
+	}()
+
+	var (
+		ackOffset int64
+		open      bool
+	)
+	chunker := &chunkWriter{
+		SessionID: msg.Header.SessionID,
+		W:         w,
+	}
+
+	buf := make([]byte, BufSize)
+
+	for {
+		windowBytes := chunker.Offset - ackOffset + ACKSlidingWindowRecv*BufSize
+		N, err := io.CopyBuffer(chunker, io.LimitReader(fd, windowBytes), buf)
+		if err != nil {
+			err = errors.Wrap(err, "failed to copy file chunk to session")
+			return err
+		}
+		if N < windowBytes {
+			break
+		}
+
+		select {
+		case msg, open = <-h.msgChan:
+			if !open {
+				return nil
+			}
+			if msg.Header.MsgType != wsft.MessageTypeACK {
+				// This should NEVER happen
+				err = errors.New("internal error")
+				return err
+			}
+			if off, ok := msg.Header.Properties["offset"].(int64); ok {
+				ackOffset = off
+			} else {
+				ackOffset += BufSize
+			}
+		case <-time.After(time.Second * 10):
+			err = errors.New("timed out waiting for ACK message")
+			return err
```

Comment:
```suggestion
			return errors.New("timed out waiting for ACK message")
```
and perhaps log with Error?